ingest-app/src/cmr/ingest/api/subscriptions.clj

```diff
@@ -124,20 +124,57 @@
       csk/->snake_case
       (str "_" (UUID/randomUUID))))
 
-(defn ingest-subscription
-  "Processes a request to create or update a subscription. Note, this will allow
-  unlimited subscriptions to be created by users for themselves. Revisit if this
-  Becomes a problem"
+(defn native-id-collision?
+  "Queries metadata db for a matching provider-id and native-id pair."
+  [context provider-id native-id]
+  (let [query {:provider-id provider-id
+               :native-id native-id
+               :exclude-metadata true
+               :latest true}]
+    (-> context
+        (mdb/find-concepts query :subscription)
+        seq)))
+
+(defn get-unique-native-id
+  "Get a native-id that is unique by testing against the database."
+  [context subscription]
+  (let [native-id (generate-native-id subscription)
+        provider-id (:provider-id subscription)]
+    (if (native-id-collision? context provider-id native-id)
+      (do
+        (warn (format "Collision detected while generating native-id [%s] for provider [%s], retrying."
+                      native-id provider-id))
+        (get-unique-native-id context subscription))
+      native-id)))
+
+(defn create-subscription
```

Comment:
I think it would be cleaner if you separate the case of ingest with native-id from ingest without native-id, i.e. don't mix create subscription with native-id and without native-id in one function.